hbase-client/src/main/java/org/apache/hadoop/hbase/zookeeper/ZKAssign.java

```diff
@@ -868,7 +868,15 @@ public static int transitionNode(ZooKeeperWatcher zkw, HRegionInfo region,
     try {
       rt = RegionTransition.createRegionTransition(
           endState, region.getRegionName(), serverName, payload);
-      if(!ZKUtil.setData(zkw, node, rt.toByteArray(), stat.getVersion())) {
+      boolean isDataSet;
+      try {
+        isDataSet = ZKUtil.setData(zkw, node, rt.toByteArray(), stat.getVersion());
+      } catch (KeeperException.BadVersionException e) {
```

Comment:
Looking at all the callers of this, it appears to me that BadVersionException needs to be handled in setData(). Currently every call to this has this pattern,

try {
setData()
} catch (BadVersionException e) {
}

Instead handle it in setData() and refactor the callers to read the return val? (on a related note, the javadoc for setData() seems incorrect. (===> If their is a version mismatch, method returns null).

I don't have much context here but it seems reasonable to me that setData() should gracefully handle version mismatches.

Reply:
Thanks @bharathv 
Agree to all points, however the reason why this Exception is not handled in setData() is because we don't want to touch other workflows, i.e. if some workflow already has some logic when they catch this Exception(many workflows do catch it), they would never get to do it anymore.
I understand that handling BVE in setData() would be better fix but there are so many workflows currently using it that we should not risk impacting them while fixing our specific case of node transitioning, but yes this can be done in another Jira given that the impact is huge and involves multiple workflows.

Reply:
That is correct, the intention is to limit the scope because there are too many callers to setData() and some seem to have their own way of handling it which we don't want to impact

Reply:
>  the intention is to limit the scope because there are too many callers to setData() and some seem to have their own way of handling it which we don't want to impact

+1

Reply:
BadVersionException signals write-write conflict on the znode. Callers have to handle it. Down in setData we have no idea what they'd want to do.